test/testdata/infer/control_flow/module_less_than.rb

```diff
@@ -10,7 +10,114 @@ def self.foo(x)
 sig {params(y: Object).void}
 def main(y)
   if y.is_a?(Module) && y < A
-    y.foo("bar") 
+    y.foo("bar")
   end
 end
 
+module M; end
+
+sig { params(obj: T.untyped).void }
+def valid?(obj)
+  T.reveal_type(obj) # error: `T.untyped`
+  if obj.is_a?(Class)
+    T.reveal_type(obj) # error: `T::Class[T.anything]`
+
+    if obj <= T.unsafe(Integer)
+      T.reveal_type(obj) # error: `T::Class[T.anything]`
+    end
+
+    if obj <= Integer
+      T.reveal_type(obj) # error: `T.class_of(Integer)`
+    end
+
+    if obj <= M
+      T.reveal_type(obj) # error: `T::Class[M]`
+    end
+  end
+
+  T.reveal_type(obj) # error: `T.untyped`
+  if obj.is_a?(Module)
+    T.reveal_type(obj) # error: `Module`
+
+    if obj <= T.unsafe(Integer)
+      T.reveal_type(obj) # error: `Module`
+    end
+
+    if obj <= Integer
+      T.reveal_type(obj) # error: `T.class_of(Integer)`
+    end
+
+    if obj <= M
+      T.reveal_type(obj) # error: `Module`
+    end
+  end
+end
+
+module Exportable
+  extend T::Sig
+  sig { returns(Integer) }
+  def export
+    0
+  end
+end
+
+sig { params(klass: T.class_of(T::Enum)).void }
+def class_lt_module(klass)
+  if klass < Exportable
+    T.reveal_type(klass) # error: `T.class_of(T::Enum)[T.all(T::Enum, Exportable)]`
+    value = klass.try_deserialize("foo")
+    raise unless value
+    exported = value.export
+    T.reveal_type(exported) # error: `Integer`
+  end
+end
+
+module FromHash
+  extend T::Sig, T::Generic
+  has_attached_class!
+  abstract!
+
+  sig { abstract.returns(T.attached_class) }
+  def from_hash
+  end
+end
+
+sig { params(klass: T.all(Module, FromHash[T.anything])).void }
+def module_lt_module(klass)
+  if klass < Exportable
+    T.reveal_type(klass) # error: `T.all(FromHash[T.anything], Module)`
+    value = klass.from_hash
+    T.reveal_type(value) # error: `T.anything`
+    raise unless value
+    exported = value.export # error: Method `export` does not exist on `T.anything`
+    T.reveal_type(exported) # error: `T.untyped`
+  end
+end
+
+module HasClassMethods
+  extend T::Helpers
+
+  module ClassMethods
+  end
+  mixes_in_class_methods(ClassMethods)
+end
+
+sig { params(klass: T::Class[T.anything], mod: Module).void }
+def nothing_special_for_class_methods(klass, mod)
+  if klass < HasClassMethods
+    T.reveal_type(klass) # error: `T::Class[HasClassMethods]`
+  end
+
+  if mod < HasClassMethods
+    T.reveal_type(mod) # error: `Module`
+  end
+end
+
+sig { params(other: T.class_of(A)).void }
+def module_should_not_add_noTypeTest(other)
+  if other <= M
+    T.reveal_type(other) # error: `T.class_of(A)[T.all(A, M)]`
+  end
+
+  T.reveal_type(other) # error: `T.class_of(A)`
```

Comment:
2 questions:

Should this be in an `else` branch, since it's testing the case where the type test is false?

For my own understanding, I commented out the `canAddNoTypeTest = false;` on line 710 in `environment.cc`. After that, this line ended up saying `Revealed type: T.class_of(A)[T.all(A, M)]`, which was surprising to me. Why would adding a `noTypeTest` lead to the knowledge that `other` is a subtype of `M` outside the `if`?

Reply:
good catch. the code which inspired this test case in Stripe's codebase had a `return` in the `if` body

What happens is that adding a `noTypeTest` makes the `else` branch dead (or, if there were a `return` in the `if` branch: makes the rest of the method dead)

So when there is both a yes and a no type test, the else branch is implicitly dead, and when sorbet joins control flow after the `else`, it says that either `other` has type `T.noreturn` or it has the type it had inside the `if`, which was `T.all(A, M)`. Since unioning anything with `T.noreturn` returns the other thing, it had the effect of changing the type even outside the `if`.

It's more explicitly to do this with a `return` in the body, so I'll add that.